src/renderer/atlas/AtlasEngine.cpp

```diff
@@ -540,1113 +479,495 @@ CATCH_RETURN()
+    _p.colorBitmap = Buffer<u32, 32>(_p.colorBitmapDepthStride * 2);
+    _p.backgroundBitmap = { _p.colorBitmap.data(), _p.colorBitmapDepthStride };
+    _p.foregroundBitmap = { _p.colorBitmap.data() + _p.colorBitmapDepthStride, _p.colorBitmapDepthStride };
+
+    memset(_p.colorBitmap.data(), 0, _p.colorBitmap.size() * sizeof(u32));
+
+    auto it = _p.unorderedRows.data();
+    for (auto& r : _p.rows)
+    {
+        r = it++;
     }
+}
 
-#ifndef NDEBUG
-    // D3D debug messages
-    if (deviceFlags & D3D11_CREATE_DEVICE_DEBUG)
+void AtlasEngine::_flushBufferLine()
+{
+    if (_api.bufferLine.empty())
     {
-        const auto infoQueue = _r.device.query<ID3D11InfoQueue>();
-        for (const auto severity : std::array{ D3D11_MESSAGE_SEVERITY_CORRUPTION, D3D11_MESSAGE_SEVERITY_ERROR, D3D11_MESSAGE_SEVERITY_WARNING })
-        {
-            infoQueue->SetBreakOnSeverity(severity, true);
-        }
+        return;
     }
-#endif // NDEBUG
 
-    {
-        wil::com_ptr<IDXGIAdapter1> dxgiAdapter;
-        THROW_IF_FAILED(_r.device.query<IDXGIObject>()->GetParent(__uuidof(dxgiAdapter), dxgiAdapter.put_void()));
-        THROW_IF_FAILED(dxgiAdapter->GetParent(__uuidof(_r.dxgiFactory), _r.dxgiFactory.put_void()));
+    const auto cleanup = wil::scope_exit([this]() noexcept {
+        _api.bufferLine.clear();
+        _api.bufferLineColumn.clear();
+    });
 
-        DXGI_ADAPTER_DESC1 desc;
-        THROW_IF_FAILED(dxgiAdapter->GetDesc1(&desc));
-        _r.d2dMode = debugForceD2DMode || WI_IsAnyFlagSet(desc.Flags, DXGI_ADAPTER_FLAG_REMOTE | DXGI_ADAPTER_FLAG_SOFTWARE);
-    }
+    // This would seriously blow us up otherwise.
+    Expects(_api.bufferLineColumn.size() == _api.bufferLine.size() + 1);
 
-    const auto featureLevel = _r.device->GetFeatureLevel();
+    auto& row = *_p.rows[_api.lastPaintBufferLineCoord.y];
 
-    if (featureLevel < D3D_FEATURE_LEVEL_10_0)
-    {
-        _r.d2dMode = true;
-    }
-    else if (featureLevel < D3D_FEATURE_LEVEL_11_0)
-    {
-        D3D11_FEATURE_DATA_D3D10_X_HARDWARE_OPTIONS options;
-        THROW_IF_FAILED(_r.device->CheckFeatureSupport(D3D11_FEATURE_D3D10_X_HARDWARE_OPTIONS, &options, sizeof(options)));
-        if (!options.ComputeShaders_Plus_RawAndStructuredBuffers_Via_Shader_4_x)
-        {
-            _r.d2dMode = true;
-        }
-    }
+    wil::com_ptr<IDWriteFontFace2> mappedFontFace;
 
-    if (!_r.d2dMode)
+#pragma warning(suppress : 26494) // Variable 'mappedEnd' is uninitialized. Always initialize an object (type.5).
+    for (u32 idx = 0, mappedEnd; idx < _api.bufferLine.size(); idx = mappedEnd)
     {
-        // Our constant buffer will never get resized
+        u32 mappedLength = 0;
+        _mapCharacters(_api.bufferLine.data() + idx, gsl::narrow_cast<u32>(_api.bufferLine.size()) - idx, &mappedLength, mappedFontFace.put());
+        mappedEnd = idx + mappedLength;
+
+        if (!mappedFontFace)

```

Comment:
To do per-glyph fallback you'd use [`IDWriteFontFallbackBuilder::AddMapping`](https://learn.microsoft.com/en-us/windows/win32/api/dwrite_2/nf-dwrite_2-idwritefontfallbackbuilder-addmapping) and use the resulting custom `IDWriteFontFallback` instead of the current one we get through `IDWriteFactory2::GetSystemFontFallback`. In other words, we can easily add per-codepoint fallback in the future!